File: pc/lib/crc32.c
#include <stdio.h>

unsigned int cal_crc32(unsigned int *ptr, unsigned int len)
{
    unsigned int xbit;
    unsigned int data;
    unsigned int crc = 0xFFFFFFFF;    /* 复位值全1 */
    unsigned int dwPolynomial = 0x04c11db7;
    while (len--)
    {
        xbit = 1 << 31;

        data = *ptr++;
        for (int bits = 0; bits < 32; bits++) 
        {
            if (crc & 0x80000000) 
            {
                crc <<= 1;
                crc ^= dwPolynomial;
            }
            else
            {
                crc <<= 1;
            }

            if (data & xbit)
            {
                crc ^= dwPolynomial;
            }

            xbit >>= 1;
        }
    }

    return crc;
}

int main(void)
{
    /* 小端(逆序) */
    unsigned int data[] = 
    {
        0x00001e00,
        0x00000e00,
        0x98fff609,
        0xc83ea800,
        0x0e00f7ff,
        0xffa90000,
        0xff99ffb1
    };

    unsigned int crc32 = 0;


    crc32 = cal_crc32(data, sizeof(data) / sizeof(unsigned int));

    printf("compute crc32:%08x\r\n", crc32);
    printf("right crc32:  d5b726ea\r\n");

    return 0;
}

File: pc/lib/crc32.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys

def cal_stm32_crc32(data):
    xbit = 0
    # 复位值为全1
    crc = 0xFFFFFFFF
    polyNomial = 0x04C11DB7

    for d in data:
        xbit = 1 << 31
        for i in range(0, 32):
            if 0x80000000 & crc:
                crc = crc << 1
                crc = crc ^ polyNomial
            else:
                crc = crc << 1

            if d & xbit:
                crc = crc ^ polyNomial
            
            xbit = xbit >> 1

    # 截断 32bit
    return crc & 0xffffffff

if __name__ == '__main__': 
    data = []
    data.append(0x00001e00)
    data.append(0x00000e00)
    data.append(0x98fff609)
    data.append(0xc83ea800)
    data.append(0x0e00f7ff)
    data.append(0xffa90000)
    data.append(0xff99ffb1)

    crc32 = cal_stm32_crc32(data)
    print("计算crc32:%08x" % crc32);
    print("传输crc32:d5b726ea");

